File: c++/DFS_BFS_1.cpp
////
//// Created by Macbook Pro on 2020/12/05.
////
//
//#include <string>
//#include <vector>
//#include <iostream>
//
//using namespace std;
//int size;
//int answer=0;
//
//void dfs(vector<int> &numbers, int target, int depth, int sum){
//
//    if(depth == size){
//        if(target == sum) answer++;
//        return;
//    }
//    dfs(numbers,target,depth+1,sum+numbers[depth]);
//    dfs(numbers,target,depth+1,sum-numbers[depth]);
//}
//
//int solution(vector<int> numbers, int target) {
//    size = numbers.size();
//    dfs(numbers,target,0,0);
//
//    return answer;
//}
//int main(){
//
//    vector<int> numbers = {1,1,1,1,1};
//    int target = 3;
//    cout << solution(numbers,target);
//
//}

File: python/programmers/dfs_bfs_1.py
# https://programmers.co.kr/learn/courses/30/lessons/43165
def dfs(numbers, target, depth, sum):
    if depth == len(numbers):
        if sum == target:
            return 1
        else:
            return 0

    return dfs(numbers, target, depth+1, sum + numbers[depth])\
           + dfs(numbers, target, depth+1, sum - numbers[depth])

def solution(numbers, target):
    answer = 0

    return dfs(numbers, target, 1, numbers[0]) + dfs(numbers, target, 1, -numbers[0])




if __name__ == '__main__':
    numbers = [1,1,1,1,1]
    target = 3
    print(solution(numbers, target))
